codex/stores/sqlitestore.nim

```diff
@@ -0,0 +1,204 @@
+## Nim-Codex
+## Copyright (c) 2021 Status Research & Development GmbH
+## Licensed under either of
+##  * Apache License, version 2.0, ([LICENSE-APACHE](LICENSE-APACHE))
+##  * MIT license ([LICENSE-MIT](LICENSE-MIT))
+## at your option.
+## This file may not be copied, modified, or distributed except according to
+## those terms.
+
+import pkg/upraises
+
+push: {.upraises: [].}
+
+import std/options
+
+import pkg/chronos
+import pkg/chronicles
+import pkg/datastore/sqlite_datastore
+import pkg/libp2p
+import pkg/questionable
+import pkg/questionable/results
+
+import ./blockstore
+import ./cachestore
+
+export blockstore, sqlite_datastore
+
+logScope:
+  topics = "codex sqlitestore"
+
+type
+  SQLiteStore* = ref object of BlockStore
+    cache: BlockStore
+    datastore: SQLiteDatastore
+
+const
+  allBlocks = Query.init(Key.init("*").expect("should not fail"))
```

Comment:
I'm really not a fan of this type of error handling, with `expect`, this is extremely fragile.

Reply:
@dryajov what you propose to use instead? Michael repalced `expect` with explicit rasising of Defect, which is the same action, just with longer code. Overall, we have a lot of `raise Defect` in nim-datastore package since we don't yet decided how to deal with all those situations

May be keep it awhile so we can start performance testing, and return back in a few weeks?

Reply:
I've removed the usage of `except` I introduced in this module and in the tests. In the case of `proc new*(T: type SQLiteStore, ...): T`, I now raise the error if construction of `SQLiteDatastore` fails.

https://github.com/status-im/nim-codex/blob/3ddcb3b9f422cd663fe3ee3d0c23753dd76f1f7f/codex/stores/sqlitestore.nim#L46-L50

We could instead make the return type `?!T` but then result handling will ripple outward, working against this initial version of the `SQLiteStore` backend being a drop-in replacement for `FSSstore`. However, I'll be happy to make those additional changes if that's preferred.

Reply:
> explicit rasising of Defect, which is the same action

It is the same action effectively, but a couple of things to consider:

Only in the case of defining the constant `allBlocks` and in `SQLiteStore` construction do I raise. In both cases that results in the original error message being displayed (the former at compile-time, latter at run-time) instead of the less helpful informative `should not fail` message.

In the other cases I added additional use of `trace` and return any failure for `Key.init`, also with the goal of making it more clear what went wrong.